src/bosh-dns/dns/server/handlers/failover_recursor_pool.go

```diff
@@ -89,26 +111,48 @@ func newSmartFailoverRecursorPool(recursors []string, logger logger.Logger) Recu
 		preferredRecursorIndex: 0,
 		logger:                 logger,
 		logTag:                 logTag,
+		recursorRetrySettings:  recursorSettings,
 	}
 }
 
 func (q *serialFailoverRecursorPool) PerformStrategically(work func(string) error) error {
 	for _, r := range q.recursors {
-		if err := work(r); err == nil {
+		if err := performRetryLogic(work, r, q.recursorRetrySettings.retryCount, q.logTag, q.logger); err == nil {
 			return nil
 		}
 	}
 
 	return ErrNoRecursorResponse
 }
 
+func performRetryLogic(work func(string) error, recursor string, retryCount int, logTag string, log logger.Logger) (err error) {
+	if retryCount == 0 {
+		err = work(recursor)
+		return err
+	}
+	for ret := 0; ret < retryCount; ret++ {
+		err = work(recursor)
+		if err == nil {
+			return err
+		}
+		if _, ok := err.(net.Error); !ok {
+			log.Debug(logTag, fmt.Sprintf("dns request error %v no retry - %v\n", err, recursor))
+			return err
+		}
+
+		log.Error(logTag, fmt.Sprintf("dns request network error %s retry [%d/%d] for recoursor %s \n", err.(net.Error), ret+1, retryCount, recursor))
+	}
```

Comment:
The retry loop will be executed only once for ` retryCount=1`.  That means we will execute only one DNS resolution and will not have two DNS resolutions (the normal one + one retry). Or am I missing something here? 

I think it makes more sense to have the firs if statement `if retryCount == 0`  integrated in the `retry` loop , so that the retry loop checks something like `ret < retryCount + 1`. In this way we will have also a fix for the comment above. WDYT?

Reply:
Hi @beyhan,

sorry for the delay.
Thx for the feedback I will check that and update the code.

There was already a Test for the retry logic you have to check new forward_handler_test file.